### PR TITLE
Validate that fields referenced in db policy for create operation are present in request body

### PR DESCRIPTION
**Pull request by ayush3797**
## Why make this change?
Fix https://github.com/Azure/data-api-builder/issues/1388

## What is this change?
A set of fields (`FieldsReferencedInDbPolicyForCreateAction`) is added to `BaseSqlQueryStructure` class which will store all the fields referenced in the db policy for create action. As we add columns to the `InsertColumns` list for `SqlInsertQueryStructure` (PUT/PATCH/POST operations) one by one, we remove the columns if present from the `FieldsReferencedInDbPolicyForCreateAction`. So, at the end if the size of this set `FieldsReferencedInDbPolicyForCreateAction` is 0, it means that all the fields referenced in the database policy for create action are being included in the insert statement, and we are good. However, if the size is non-zero, we throw a bad request exception.

## How was this tested?

- [x] Integration Tests

## Sample Request(s)

**Operation Type**: POST
**Api**: REST
**URL**: https://localhost:5001/api/Revenue/
**Database policy configured for operation:** "@item.revenue gt 1000"
**Reason why we will get an exception:** The field `revenue` referenced in the database policy is not one amongst the fields that are being inserted.

Request body: 
```json
{
    "id": 18,
    "category":"book",
    "accessible_role": "Anonymous"
}
```

Response:
```json
{
    "error": {
        "code": "BadRequest",
        "message": "One or more fields referenced by the database policy are not present in the request body.",
        "status": 400
    }
}
```

### CODE REVIEW

File: src/Service/Resolvers/BaseSqlQueryBuilder.cs [hunk 1]
```diff
@@ -361,7 +361,9 @@ protected string Build(List<string> columns)
         /// <summary>
         /// Join predicate strings while ignoring empty or null predicates
         /// </summary>
-        /// <returns>returns "1 = 1" if no valid predicates</returns>
+        /// <param name="predicateStrings">Array of predicate strings coming from filters in query string,
+        /// primary key predicates for update operations or database policies.</param>
```
Comment (Aniruddh25) [inline src/Service/Resolvers/BaseSqlQueryBuilder.cs:365]: why only "update" operations?

File: src/Service/Resolvers/Sql Query Structures/SqlInsertQueryStructure.cs [hunk 2]
```diff
@@ -74,6 +74,15 @@
                 MetadataProvider.TryGetBackingColumn(EntityName, param.Key, out string? backingColumn);
                 PopulateColumnsAndParams(backingColumn!, param.Value);
             }
+
+            if (FieldsReferencedInDbPolicyForCreateAction.Count > 0)
+            {
+                // This indicates that one or more fields referenced in the database policy are not a part of the insert statement.
+                throw new DataApiBuilderException(
+                    message: "One or more fields referenced by the database policy are not present in the request body.",
```
Comment (Aniruddh25) [inline src/Service/Resolvers/Sql Query Structures/SqlInsertQueryStructure.cs:82]: Does this work for graphql? 

Have we tested create policy for graphql?

File: src/Service/Resolvers/BaseSqlQueryBuilder.cs (hunk 1 above)
Comment (ayush3797) [reply src/Service/Resolvers/BaseSqlQueryBuilder.cs:365]: Oh yeah, will update the comment as the predicates can be for other operations as well, thanks.

File: src/Service/Resolvers/Sql Query Structures/SqlInsertQueryStructure.cs (hunk 2 above)
Comment (ayush3797) [reply src/Service/Resolvers/Sql Query Structures/SqlInsertQueryStructure.cs:84]: Yes it works for graphql as well. In process of writing the test case for that.

**Review by Aniruddh25 — APPROVED** [review]
Thanks for fixing the bug timely! 
LGTM, after checking GraphQL.

**Review by abhishekkumams — APPROVED** [review]

**Pull request merged**